cloudmarker/clouds/azsql.py

```diff
@@ -0,0 +1,231 @@
+        self._max_subs = _max_subs
+        self._max_recs = _max_recs
+        _log.info('Initialized; tenant: %s; processes: %s; threads: %s',
+                  self._tenant, self._processes, self._threads)
+
+    def read(self):
+        """Return an Azure SQL database record.
+
+        Yields:
+            dict: An Azure SQL database record.
+
+        """
+        yield from ioworkers.run(self._get_tenant_dbs,
+                                 self._get_server_db_details,
+                                 self._processes, self._threads,
+                                 __name__)
+
+    def _get_tenant_dbs(self):
+        """Get SQL DBs from all subscriptions in a tenant.
+
+        The yielded tuples when unpacked would become arguments for
+        :meth:`_get_server_db_details`. Each such tuple represents a
+        single unit of work that :meth:`_get_server_db_details` can
+        work on independently in its own worker thread.
+
+        Yields:
+            tuple: A tuple which when unpacked forms valid arguments for
+                :meth:`_get_server_db_details`.
+
+        """
+        try:
+            tenant = self._tenant
+            creds = self._credentials
+            sub_client = SubscriptionClient(creds)
+            sub_list = sub_client.subscriptions.list()
+
+            for sub_index, sub in enumerate(sub_list):
+                sub = sub.as_dict()
+                _log.info('Found %s', util.outline_az_sub(sub_index,
+                                                          sub, tenant))
+
+                yield from self._get_subscription_dbs(sub_index, sub)
+                # Break after pulling data for self._max_subs number of
+                # subscriptions. Note that if self._max_subs is 0 or less,
+                # then the following condition never evaluates to True.
+                if sub_index + 1 == self._max_subs:
+                    _log.info('Stopping subscriptions fetch due to '
+                              '_max_subs: %d; tenant: %s', self._max_subs,
+                              tenant)
+                    break
+
+        except Exception as e:
+            _log.error('Failed to fetch subscriptions; %s; error: %s: %s',
+                       util.outline_az_sub(sub_index, sub, tenant),
+                       type(e).__name__, e)
+
+    def _get_subscription_dbs(self, sub_index, sub):
+        """Get SQL DBs from a single subscrption.
+
+        Yields:
+            tuple: A tuple which when unpacked forms valid arguments for
+                :meth:`_get_server_db_details`.
+
+        """
+        try:
+            tenant = self._tenant
+            creds = self._credentials
+            sub_id = sub.get('subscription_id')
+            sql_client = SqlManagementClient(creds, sub_id)
+            db_server_list = sql_client.servers.list()
+
+            for server_index, sql_server in enumerate(db_server_list):
+                sql_server = sql_server.as_dict()
```

Comment:
This `sql_server` response that was received is not being yielded by this plugin so that it could be saved in the stores. I think this is useful information we are discarding. Do you have any thoughts around saving this information? It does not have to be in this pull request but I just want to know your thoughts regarding this.

Reply:
I gave it some though and have plan to add it to `AzCloud` plugin. `AzCloud` retrieves data about `MySQL` and `PostgreSQL` servers. I plan to add `SQL` sever there for consistency.

Reply:
Yes, that sounds good.